OpenRA.Game/Support/ConditionExpression.cs

```diff
@@ -0,0 +1,898 @@
+									lastToken.Symbol, lastToken.Index, token.Symbol, token.Index));
+						throw new InvalidDataException("Missing binary operation before `{0}` at index {1}".F(token.Symbol, token.Index));
+					}
+				}
+
+				if (token.Type == TokenType.Variable)
+					variables.Add(token.Symbol);
+
+				tokens.Add(token);
+				lastToken = token;
+			}
+
+			if (currentOpeners.Count > 0)
+				throw new InvalidDataException("Unclosed opening parenthesis at index {0}".F(currentOpeners.Peek().Index));
+
+			asFunction = new Compiler().Compile(ToPostfix(tokens).ToArray());
+		}
+
+		static int ParseSymbol(string symbol, IReadOnlyDictionary<string, int> symbols)
+		{
+			int value;
+			symbols.TryGetValue(symbol, out value);
+			return value;
+		}
+
+		static IEnumerable<Token> ToPostfix(IEnumerable<Token> tokens)
+		{
+			var s = new Stack<Token>();
+			foreach (var t in tokens)
+			{
+				if (t.Opens != Grouping.None)
+					s.Push(t);
+				else if (t.Closes != Grouping.None)
+				{
+					Token temp;
+					while (!((temp = s.Pop()).Opens != Grouping.None))
+						yield return temp;
+				}
+				else if (t.OperandSides == OperandSides.None)
+					yield return t;
+				else
+				{
+					while (s.Count > 0 && ((t.Associativity == Associativity.Right && t.Precedence < s.Peek().Precedence)
+						|| (t.Associativity == Associativity.Left && t.Precedence <= s.Peek().Precedence)))
+						yield return s.Pop();
+
+					s.Push(t);
+				}
+			}
+
+			while (s.Count > 0)
+				yield return s.Pop();
+		}
+
+		enum ExpressionType { Int, Bool }
+
+		static readonly ParameterExpression SymbolsParam =
+			Expressions.Expression.Parameter(typeof(IReadOnlyDictionary<string, int>), "symbols");
+		static readonly ConstantExpression Zero = Expressions.Expression.Constant(0);
+		static readonly ConstantExpression One = Expressions.Expression.Constant(1);
+		static readonly ConstantExpression False = Expressions.Expression.Constant(false);
+		static readonly ConstantExpression True = Expressions.Expression.Constant(true);
+
+		static Expression AsBool(Expression expression)
+		{
+			return Expressions.Expression.GreaterThan(expression, Zero);
+		}
+
+		static Expression AsNegBool(Expression expression)
+		{
+			return Expressions.Expression.LessThanOrEqual(expression, Zero);
+		}
+
+		static Expression IfThenElse(Expression test, Expression ifTrue, Expression ifFalse)
+		{
+			return Expressions.Expression.Condition(test, ifTrue, ifFalse);
+		}
+
+		class AstStack
+		{
+			readonly List<Expression> expressions = new List<Expression>();
+			readonly List<ExpressionType> types = new List<ExpressionType>();
+
+			public ExpressionType PeekType() { return types[types.Count - 1]; }
+
+			public Expression Peek(ExpressionType toType)
+			{
+				var fromType = types[types.Count - 1];
+				var expression = expressions[expressions.Count - 1];
+				if (toType == fromType)
+					return expression;
+
+				switch (fromType)
```

Comment:
Minor style nit: This would be clearer if you switched on `toType` instead.  I had to stop and read this method carefully when I noticed that the `Bool` case was returning ints and visa versa.